pvlib/iotools/bsrn.py

```diff
@@ -4,139 +4,449 @@
+        # Change to station sub-directory (checks that the station exists)
+        try:
+            ftp.cwd(f'/{station}')
+        except ftplib.error_perm as e:
+            raise KeyError('Station sub-directory does not exist. Specified '
+                           'station is probably not a proper three letter '
+                           'station abbreviation.') from e
+        dfs = []  # Initialize list for monthly dataframes
+        non_existing_files = []  # Initilize list of files that were not found
+        for filename in filenames:
+            try:
+                bio = io.BytesIO()  # Initialize BytesIO object
+                # Retrieve binary file from server and write to BytesIO object
+                response = ftp.retrbinary(f'RETR {filename}', bio.write)
+                # Check that transfer was successfull
+                if not response.startswith('226 Transfer complete'):
+                    raise ftplib.Error(response)
+                # Save file locally if local_path is specified
+                if local_path is not None:
+                    # Create local file
+                    with open(os.path.join(local_path, filename), 'wb') as f:
+                        f.write(bio.getbuffer())  # Write local file
+                # Decompress/unzip and decode the binary file
+                text = gzip.decompress(bio.getvalue()).decode('latin1')
+                # Convert string to StringIO and parse data
+                dfi, metadata = parse_bsrn(io.StringIO(text), logical_records)
+                dfs.append(dfi)
+            # FTP client raises an error if the file does not exist on server
+            except ftplib.error_perm as e:
+                if str(e) == '550 Failed to open file.':
+                    non_existing_files.append(filename)
+                else:
+                    raise ftplib.error_perm(e)
+        ftp.quit()  # Close and exit FTP connection
 
+    # Raise user warnings
+    if not dfs:  # If no files were found
+        warnings.warn('No files were available for the specified timeframe.')
+    elif non_existing_files:  # If only some files were missing
+        warnings.warn(f'The following files were not found: {non_existing_files}')  # noqa: E501
+
+    # Concatenate monthly dataframes to one dataframe
+    if len(dfs):
+        data = pd.concat(dfs, axis='rows')
+    else:  # Return empty dataframe
+        data = _empty_dataframe_from_logical_records(logical_records)
+        metadata = {}
+    # Return dataframe and metadata (metadata belongs to last available file)
+    return data, metadata
+
+
+def parse_bsrn(fbuf, logical_records=['0100']):
+    """
+    Parse a file-like buffer of a BSRN station-to-archive file.
 
     Parameters
     ----------
-    filename: str
-        A relative or absolute file path.
+    fbuf: file-like buffer
+        Buffer of a BSRN station-to-archive data file
+    logical_records: str or list, default: ['0100']

```

Comment:
Maybe I'm doing it wrong but I don't think `str` is actually allowed here (or the other functions)

Reply:
In the parse function I have this line, which allows it to be string:
`    for lr in list(logical_records):`

This does however also have to be added to the empty dataframe function.

First though, is this a desireable functionality or should I remove this?

Reply:
Note that `list('0100') == ['0', '1', '0', '0']`.  Does it work for you if you pass `logical_records='0100'`?  I get `ValueError: Logical record 0 not in ['0100', '0300', '0500'].`  I'd say just require it to be a list (or tuple etc).

Reply:
You are indeed correct! List or tuples it is :)